enable unit tests

deepakrkris opened the pull request:
<!--
Before submitting a pull request, please read
https://github.com/nodejs/node-addon-api/blob/main/CONTRIBUTING.md.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

deepakrkris commented:
@mhdawson @KevinEady @gabrielschulhof PR is ready for review

gabrielschulhof commented on test/index.js:
```diff
@@ -29,9 +29,23 @@ if (typeof global.gc !== 'function') {
 
 const fs = require('fs');
 const path = require('path');
+process.env.filter = require('../unit-test/matchModules').matchWildCards(process.env.filter);
```
This could easily be process.env.npm_config_filter.

gabrielschulhof commented on unit-test/binding.gyp:
```diff
@@ -0,0 +1,64 @@
+      'dependencies': [ 'generateBindingCC' ]
+    },
+  ],
+}
```
Please add a newline!

gabrielschulhof commented on unit-test/binding.gyp:
```diff
@@ -0,0 +1,64 @@
+{
+  'target_defaults': {
+    'includes': ['common.gypi'],
```
Why not ../common.gypi and then you don't need to copy the file?

gabrielschulhof commented on unit-test/exceptions.js:
```diff
@@ -0,0 +1,34 @@
+/**
+ * 
+ * This file points out anamolies/exceptions in test files when generating the binding.cc file
```
anomalies

gabrielschulhof commented on unit-test/generate-binding-cc.js:
```diff
@@ -0,0 +1,69 @@
+generateBindingConfigurations().then(generateFileContent).then(writeToBindingFile);
+
+/**
+ *
+ * Test cases
+ * @fires only when run directly from terminal with TEST=true
+ *
+ *
+ *
+ * 
```
Do we need these blank lines?

gabrielschulhof replied:
Do we need the blank lines surrounding the comment?

gabrielschulhof commented on unit-test/generate-binding-cc.js:
```diff
@@ -0,0 +1,69 @@
+generateBindingConfigurations().then(generateFileContent).then(writeToBindingFile);
+
+/**
+ *
```
Do we need this blank line?

gabrielschulhof commented on unit-test/injectTestParams.js:
```diff
@@ -0,0 +1,106 @@
+  fs.mkdirSync('./generated/');
+
+/**
+ *
+ *  @returns : list of files to compile by node-gyp
+ *  @param : none
+ *  @requires : picks `filter` parameter from process.env
+ *  This function is used as an utility method to inject a list of files to compile into binding.gyp
+ *
+ *
```
Same here. Do we need the first and trailing blank lines?

gabrielschulhof commented on unit-test/injectTestParams.js:
```diff
@@ -0,0 +1,106 @@
+ * @param : none
+ * @requires : picks `filter` parameter from process.env
+ * This function is used as an utility method by the generateBindingCC step in binding.gyp
+ *
```
Blank line.

gabrielschulhof commented on unit-test/injectTestParams.js:
```diff
@@ -0,0 +1,106 @@
+ *
+ * Test cases
+ * @fires only when run directly from terminal
+ *
+ *
+ *
```
Do we need these blank lines? Here and in all these comments?

gabrielschulhof commented on unit-test/spawnTask.js:
```diff
@@ -0,0 +1,26 @@
+/*
+*  spawns a child process to run a given node.js script
+*/
+module.exports.runChildProcess = async function (scriptName, options) {
```
The function doesn't need to be async if it returns a promise.

gabrielschulhof replied:
It can be used from an async function without itself being marked as async.

gabrielschulhof commented on test/common/index.js:
```diff
@@ -64,50 +63,50 @@ function _mustCallInner(fn, criteria, field) {
-exports.mustNotCall = function(msg) {
-  return function mustNotCall() {
+exports.mustNotCall = function (msg) {
+  return function mustNotCall () {
```
Yes, it's the result of running the linter ☺

deepakrkris commented:
@gabrielschulhof I have changed the npm_config_filter refererence now.

https://github.com/nodejs/node-addon-api/pull/1078#discussion_r739338275

gabrielschulhof commented on package.json:
```diff
@@ -378,7 +378,9 @@
     "lint": "eslint $(git diff --name-only refs/remotes/origin/main '**/*.js' | xargs) && node tools/clang-format",
-    "lint:fix": "node tools/clang-format --fix && eslint --fix $(git diff --cached --name-only '**/*.js' | xargs && git diff --name-only '**/*.js' | xargs)"
+    "lint:fix": "node tools/clang-format --fix && eslint --fix $(git diff --cached --name-only '**/*.js' | xargs && git diff --name-only '**/*.js' | xargs)",
+    "preunit": "filter=\"$npm_config_filter\" node-gyp rebuild -C unit-test",
```
You can drop `"filter=\"$npm_config_filter\"` because you're using `process.env.npm_config_filter` inside the script now.

KevinEady commented:
So, some of the addons don't run correctly in unit-test mode. I think we can tackle that at a later date. However, I think there is a bigger issue in that a failed unit test exits the `npm run unit --filter=<filter>` process _successfully_:

```
➜  node-addon-api git:(b398038) ✗ npm run unit --filter=addon ; echo "Return code: $?"


> node-addon-api@4.2.0 preunit /Users/kevineady/Documents/Projects/node-addon-api
> filter="$npm_config_filter" node-gyp rebuild -C unit-test

  ACTION Generating binding cc file generated/binding.cc
test modules to bind:  [ 'addon' ]
generated binding file  /Users/kevineady/Documents/Projects/node-addon-api/unit-test/generated/binding.cc 2021-12-03T15:40:16.308Z
  TOUCH Release/obj.target/generateBindingCC.stamp
  CXX(target) Release/obj.target/binding/generated/binding.o
  CXX(target) Release/obj.target/binding/../test/addon.o
  SOLINK_MODULE(target) Release/binding.node
  CXX(target) Release/obj.target/binding_noexcept/generated/binding.o
  CXX(target) Release/obj.target/binding_noexcept/../test/addon.o
  SOLINK_MODULE(target) Release/binding_noexcept.node
  CXX(target) Release/obj.target/binding_noexcept_maybe/generated/binding.o
  CXX(target) Release/obj.target/binding_noexcept_maybe/../test/addon.o
  SOLINK_MODULE(target) Release/binding_noexcept_maybe.node
  CXX(target) Release/obj.target/binding_swallowexcept/generated/binding.o
  CXX(target) Release/obj.target/binding_swallowexcept/../test/addon.o
  SOLINK_MODULE(target) Release/binding_swallowexcept.node
  CXX(target) Release/obj.target/binding_swallowexcept_noexcept/generated/binding.o
  CXX(target) Release/obj.target/binding_swallowexcept_noexcept/../test/addon.o
  SOLINK_MODULE(target) Release/binding_swallowexcept_noexcept.node

> node-addon-api@4.2.0 unit /Users/kevineady/Documents/Projects/node-addon-api
> filter="$npm_config_filter" node unit-test/test

Starting to run tests in  /Users/kevineady/Documents/Projects/node-addon-api/unit-test/unit-test 2021-12-03T15:40:22.515Z
napiVersion:5

Testing with Node-API Version '5'.

Starting test suite
 [ 'addon' ]

Running test 'addon'

error: dyld: lazy symbol binding failed: Symbol not found: __Z9InitAddonN4Napi3EnvE
  Referenced from: /Users/kevineady/Documents/Projects/node-addon-api/unit-test/build/Release/binding.node
  Expected in: flat namespace

dyld: Symbol not found: __Z9InitAddonN4Napi3EnvE
  Referenced from: /Users/kevineady/Documents/Projects/node-addon-api/unit-test
error: /build/Release/binding.node
  Expected in: flat namespace


error: Tests aborted with SIGABRT

child process exited with code 1
Completed running tests 2021-12-03T15:40:24.973Z
Return code: 0
```

I think this is a blocker for merging the PR, as the return value should signify if the test passes or not.

gabrielschulhof commented:
@deepakrkris please also remove the aliasing of npm_config_filter to filter from the command lines in package.json.

gabrielschulhof reviewed (changes requested):
Marking as `Request changes` per @KevinEady's comment about returning success even on error.

deepakrkris commented:
> @deepakrkris please also remove the aliasing of npm_config_filter to filter from the command lines in package.json.

@gabrielschulhof I have fixed this now

deepakrkris commented:
> Marking as `Request changes` per @KevinEady's comment about returning success even on error. Please dismiss when the problem is fixed!

@KevinEady @gabrielschulhof I have fixed this now

deepakrkris commented:
@KevinEady I am able verify the fix as below

```
SFOC02CC4L3MD6T:node-addon-api t_rajad$ npm run unit --filter=addon ; echo "Return code: $?"

> node-addon-api@4.2.0 preunit /Users/t_rajad/projects/AA-NAPI/node-addon-api
> node-gyp rebuild -C unit-test

  ACTION Generating binding cc file generated/binding.cc
test modules to bind:  [ 'addon' ]
generated binding file  /Users/t_rajad/projects/AA-NAPI/node-addon-api/unit-test/generated/binding.cc 2021-12-10T15:48:28.973Z
  TOUCH Release/obj.target/generateBindingCC.stamp
  CXX(target) Release/obj.target/binding/generated/binding.o
  CXX(target) Release/obj.target/binding/../test/addon.o
  SOLINK_MODULE(target) Release/binding.node
  CXX(target) Release/obj.target/binding_noexcept/generated/binding.o
  CXX(target) Release/obj.target/binding_noexcept/../test/addon.o
  SOLINK_MODULE(target) Release/binding_noexcept.node
  CXX(target) Release/obj.target/binding_noexcept_maybe/generated/binding.o
  CXX(target) Release/obj.target/binding_noexcept_maybe/../test/addon.o
  SOLINK_MODULE(target) Release/binding_noexcept_maybe.node
  CXX(target) Release/obj.target/binding_swallowexcept/generated/binding.o
  CXX(target) Release/obj.target/binding_swallowexcept/../test/addon.o
  SOLINK_MODULE(target) Release/binding_swallowexcept.node
  CXX(target) Release/obj.target/binding_swallowexcept_noexcept/generated/binding.o
  CXX(target) Release/obj.target/binding_swallowexcept_noexcept/../test/addon.o
  SOLINK_MODULE(target) Release/binding_swallowexcept_noexcept.node

> node-addon-api@4.2.0 unit /Users/t_rajad/projects/AA-NAPI/node-addon-api
> node unit-test/test

Starting to run tests in  /Users/t_rajad/projects/AA-NAPI/node-addon-api/unit-test/unit-test 2021-12-10T15:48:38.701Z
napiVersion:7

Testing with Node-API Version '7'.

Starting test suite
 [ 'addon' ]
Running test 'addon'

TypeError: binding.addon.increment1 is not a function
    at test (/Users/t_rajad/projects/AA-NAPI/node-addon-api/test/addon.js:8:36)
    at Object.exports.runTest (/Users/t_rajad/projects/AA-NAPI/node-addon-api/test/common/index.js:88:27)
    at Object.<anonymous> (/Users/t_rajad/projects/AA-NAPI/node-addon-api/test/addon.js:5:38)
    at Module._compile (internal/modules/cjs/loader.js:1015:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1035:10)
    at Module.load (internal/modules/cjs/loader.js:879:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:903:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at /Users/t_rajad/projects/AA-NAPI/node-addon-api/test/index.js:150:11

child process exited with code 1
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! node-addon-api@4.2.0 unit: `node unit-test/test`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the node-addon-api@4.2.0 unit script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/t_rajad/.npm/_logs/2021-12-10T15_48_39_018Z-debug.log
Return code: 1
```

KevinEady reviewed (approved):
LGTM! We should figure out why individual unit tests fail, but I think that can come at a different time.

I have verified the return code is now non-zero for failed unit tests:

```
➜  node-addon-api git:(bc7622e) ✗ npm run unit --filter=addon ; echo "Return code: $?"


> node-addon-api@4.2.0 preunit /Users/kevineady/Documents/Projects/node-addon-api
> node-gyp rebuild -C unit-test

  ACTION Generating binding cc file generated/binding.cc
test modules to bind:  [ 'addon' ]
generated binding file  /Users/kevineady/Documents/Projects/node-addon-api/unit-test/generated/binding.cc 2021-12-10T16:03:38.620Z
  TOUCH Release/obj.target/generateBindingCC.stamp
  CXX(target) Release/obj.target/binding/generated/binding.o
  CXX(target) Release/obj.target/binding/../test/addon.o
  SOLINK_MODULE(target) Release/binding.node
  CXX(target) Release/obj.target/binding_noexcept/generated/binding.o
  CXX(target) Release/obj.target/binding_noexcept/../test/addon.o
  SOLINK_MODULE(target) Release/binding_noexcept.node
  CXX(target) Release/obj.target/binding_noexcept_maybe/generated/binding.o
  CXX(target) Release/obj.target/binding_noexcept_maybe/../test/addon.o
  SOLINK_MODULE(target) Release/binding_noexcept_maybe.node
  CXX(target) Release/obj.target/binding_swallowexcept/generated/binding.o
  CXX(target) Release/obj.target/binding_swallowexcept/../test/addon.o
  SOLINK_MODULE(target) Release/binding_swallowexcept.node
  CXX(target) Release/obj.target/binding_swallowexcept_noexcept/generated/binding.o
  CXX(target) Release/obj.target/binding_swallowexcept_noexcept/../test/addon.o
  SOLINK_MODULE(target) Release/binding_swallowexcept_noexcept.node

> node-addon-api@4.2.0 unit /Users/kevineady/Documents/Projects/node-addon-api
> node unit-test/test

Starting to run tests in  /Users/kevineady/Documents/Projects/node-addon-api/unit-test/unit-test 2021-12-10T16:03:49.692Z
napiVersion:5

Testing with Node-API Version '5'.

Starting test suite
 [ 'addon' ]

Running test 'addon'

error: dyld: lazy symbol binding failed: Symbol not found: __Z9InitAddonN4Napi3EnvE
  Referenced from: /Users/kevineady/Documents/Projects/node-addon-api/unit-test/build/Release/binding.node
  Expected in: flat namespace

dyld: Symbol not found: __Z9InitAddonN4Napi3EnvE
  Referenced from: /Users/kevineady/Documents/Projects/node-addon-api/unit-test/build/Release/binding.node
  Expected in: flat namespace


error: Tests aborted with SIGABRT

child process exited with code 1
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! node-addon-api@4.2.0 unit: `node unit-test/test`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the node-addon-api@4.2.0 unit script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/kevineady/.npm/_logs/2021-12-10T16_03_50_458Z-debug.log

Return code: 1
```

deepakrkris commented:
@KevinEady I cloned the repo (from my fork) freshly and the unit test with filter=addon runs good

```SFOC02CC4L3MD6T:node-addon-api t_rajad$ npm run unit --filter=addon

> node-addon-api@4.2.0 preunit /Users/t_rajad/projects/AA-NAPI/node-addon-api
> node-gyp rebuild -C unit-test

  ACTION Generating binding cc file generated/binding.cc
test modules to bind:  [ 'addon' ]
generated binding file  /Users/t_rajad/projects/AA-NAPI/node-addon-api/unit-test/generated/binding.cc 2021-12-17T15:54:59.530Z
  TOUCH Release/obj.target/generateBindingCC.stamp
  CXX(target) Release/obj.target/binding/generated/binding.o
  CXX(target) Release/obj.target/binding/../test/addon.o
  SOLINK_MODULE(target) Release/binding.node
  CXX(target) Release/obj.target/binding_noexcept/generated/binding.o
  CXX(target) Release/obj.target/binding_noexcept/../test/addon.o
  SOLINK_MODULE(target) Release/binding_noexcept.node
  CXX(target) Release/obj.target/binding_noexcept_maybe/generated/binding.o
  CXX(target) Release/obj.target/binding_noexcept_maybe/../test/addon.o
  SOLINK_MODULE(target) Release/binding_noexcept_maybe.node
  CXX(target) Release/obj.target/binding_swallowexcept/generated/binding.o
  CXX(target) Release/obj.target/binding_swallowexcept/../test/addon.o
  SOLINK_MODULE(target) Release/binding_swallowexcept.node
  CXX(target) Release/obj.target/binding_swallowexcept_noexcept/generated/binding.o
  CXX(target) Release/obj.target/binding_swallowexcept_noexcept/../test/addon.o
  SOLINK_MODULE(target) Release/binding_swallowexcept_noexcept.node

> node-addon-api@4.2.0 unit /Users/t_rajad/projects/AA-NAPI/node-addon-api
> node unit-test/test

Starting to run tests in  /Users/t_rajad/projects/AA-NAPI/node-addon-api/unit-test/unit-test 2021-12-17T15:55:08.468Z
napiVersion:7

Testing with Node-API Version '7'.

Starting test suite
 [ 'addon' ]
Running test 'addon'


All tests passed!

child process exited with code 0
SFOC02CC4L3MD6T:node-addon-api t_rajad$ 
```

deepakrkris commented:
@KevinEady I cloned the forked branch and worked on it. Is there another way you have cloned the repo like say cloning the root repo and then checking out to a forked branch ? I can try that as well.

gabrielschulhof commented on test/index.js:
```diff
@@ -27,17 +27,34 @@ if (typeof global.gc !== 'function') {
+let filterCondition = process.env.npm_config_filter || '';
+let filterConditionFiles = [];
+
+if (process.env.npm_config_filter !== null && process.env.npm_config_filter !== undefined && process.env.npm_config_filter.trim() !== '') {
```
Why not `if (filterCondition != null && ...)`, accounting for the fact that the default is `''` and not `null`?

deepakrkris replied:
agreed

gabrielschulhof commented on test/index.js:
```diff
@@ -27,17 +27,34 @@ if (typeof global.gc !== 'function') {
+
+if (process.env.npm_config_filter !== null && process.env.npm_config_filter !== undefined && process.env.npm_config_filter.trim() !== '') {
+  filterCondition = require('../unit-test/matchModules').matchWildCards(process.env.npm_config_filter);
+  filterConditionFiles = filterCondition.split(' ').length ? filterCondition.split(' ') : [filterCondition];
```
You should make this check explicitly `> 0`.

deepakrkris replied:
agreed

gabrielschulhof commented on test/index.js:
```diff
@@ -27,17 +27,34 @@ if (typeof global.gc !== 'function') {
+function checkFilterCondition (fileName, parsedFilepath) {
+  let result = false;
+
+  if (!filterConditionFiles.length) result = true;
```
Same here (`> 0`).

deepakrkris replied:
agreed

gabrielschulhof commented on test/index.js:
```diff
@@ -87,7 +108,7 @@ if (napiVersion < 3) {
 }
 
-if (napiVersion < 4) {
+if (napiVersion < 4 && !filterConditionFiles.length) {
```
Please don't treat `.length` as a boolean!

mhdawson reviewed (approved):
LGTM

mhdawson commented:
@deepakrkris I see there are a few final comments from @gabrielschulhof . If you can address those I'll land.

mhdawson commented:
@deepakrkris if you can squash down to 1 commit that would also make it much easier for me to land and it seems individual commits have conflicts.

mhdawson commented:
@deepakrkris ran the ci jobs https://ci.nodejs.org/job/node-test-node-addon-api-new/ for latest branch and 12.x and all seems ok so we should be good once you update the last few comments. If you can email me once you do that that will help me get to sooner.

deepakrkris commented:
> @deepakrkris I see there are a few final comments from @gabrielschulhof . If you can address those I'll land.

@mhdawson sure, I will fix the comments and ping here in a few hours, thank you

deepakrkris commented:
> @deepakrkris if you can squash down to 1 commit that would also make it much easier for me to land and it seems individual commits have conflicts.

@mhdawson I have fixed comments from Gabriel and squashed the commits
